infra/src/main/kotlin/fi/fta/geoviite/infra/publication/ValidationContext.kt

```diff
@@ -0,0 +1,399 @@
+package fi.fta.geoviite.infra.publication
+
+import fi.fta.geoviite.infra.common.*
+import fi.fta.geoviite.infra.geocoding.GeocodingContextCacheKey
+import fi.fta.geoviite.infra.geocoding.GeocodingService
+import fi.fta.geoviite.infra.switchLibrary.SwitchLibraryService
+import fi.fta.geoviite.infra.tracklayout.*
+import java.util.concurrent.ConcurrentHashMap
+
+class NullableCache<K, V> {
+    val map: ConcurrentHashMap<K, NullableValue<V>> = ConcurrentHashMap()
+
+    fun get(key: K, getter: (K) -> V?): V? = map.computeIfAbsent(key) { k -> NullableValue(getter(k)) }.value
+
+    fun preload(keys: List<K>, getter: (List<K>) -> Map<K, V?>) {
+        val missingKeys = keys.filterNot(map::containsKey)
+        val preloaded = if (missingKeys.isNotEmpty()) getter(missingKeys) else mapOf()
+        // Also cache the "not found" results
+        val notFound = missingKeys.filterNot(preloaded::containsKey)
+        putAll(preloaded + notFound.associateWith { null })
+    }
+
+    fun putAll(values: Map<K, V?>): Unit = map.putAll(values.mapValues { (_, v) -> NullableValue(v) })
+
+    fun contains(key: K): Boolean = map.containsKey(key)
+
+    // Holder for nullable value so computeIfAbsent doesn't re-resolve a missing result
+    data class NullableValue<T>(val value: T?)
+}
+
+class NameCache<Field, T : Draftable<T>>(val fetch: (List<Field>) -> Map<Field, List<IntId<T>>>) {
+    val map: ConcurrentHashMap<Field, List<IntId<T>>> = ConcurrentHashMap()
+
+    fun get(name: Field) = map.computeIfAbsent(name) { n -> fetch(listOf(n))[n] ?: emptyList() }
+
+    fun preload(names: List<Field>) = map.putAll(fetch(names))
+}
+
+typealias RowVersionCache<T> = NullableCache<IntId<T>, RowVersion<T>>
```

Comment:
Näistä en ole varma kannattaisiko edes vaivautua mutta tuntuu minusta luettavammalta kuin koko tyypin auki kirjoittaminen. Opinions?

Reply:
Vähän kumsi-kumsa tyyppinen judu. Sinänsä RowVersionCachen tyyppitiedosta on hankalaa päätellä, että kyseessä onkin hieman muokattu mappi-implementaatio, kun taas NullableCache<..., ...> sisältää kaksipaikkaisen tyyppitiedon, josta mappityypin päättely saattaa olla helpompi tehdä, mutta tyyppitieto on taas käyttöpaikoissa pidempi, joskin ei kuitenkaan kovin pitkä vielä.

OK näin tai toisin 😄

Reply:
Mielestäni nää typealiakset on hyviä. Kertovat paremmin käyttötarkoituksen, ja tuo pohjatyyppikin on aika looginen kumminkin.